src/core/collector/mod.rs

```diff
@@ -0,0 +1,757 @@
+};
+
+use codec::{Decode, Encode};
+use color_eyre::eyre::eyre;
+use subxt::{
+	ext::{
+		sp_core::{Hasher, H256},
+		sp_runtime::traits::BlakeTwo256,
+	},
+	PolkadotConfig,
+};
+use tokio::sync::{
+	broadcast::{self, Receiver as BroadcastReceiver, Sender as BroadcastSender},
+	mpsc::{error::TryRecvError, Receiver as MspcReceiver},
+};
+
+pub mod candidate_record;
+mod ws;
+
+use crate::core::{
+	ApiService, RecordTime, RecordsStorageConfig, StorageEntry, SubxtCandidateEvent, SubxtCandidateEventType,
+	SubxtDispute, SubxtDisputeResult, SubxtEvent,
+};
+
+use candidate_record::*;
+use ws::*;
+
+#[derive(Clone, Debug, Parser)]
+#[clap(rename_all = "kebab-case")]
+pub struct CollectorOptions {
+	/// Maximum blocks to store
+	#[clap(name = "max-blocks", long)]
+	max_blocks: Option<usize>,
+	/// WS listen address to bind to
+	#[clap(short = 'l', long = "listen")]
+	listen_addr: Option<SocketAddr>,
+}
+
+/// This type is used to distinguish different keys in the storage
+#[derive(Clone, Copy, Debug, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub enum CollectorPrefixType {
+	/// Candidate prefixed by Parachain-Id
+	Candidate(u32),
+	/// A mapping to find out parachain id by candidate hash (e.g. when we don't know the parachain)
+	CandidatesParachains,
+	/// Relay chain block
+	Head,
+	/// Session data, keyed by session hash (blake2b(session_index))
+	Session,
+	/// Inherent data (more expensive to store, so good to have it shared)
+	InherentData,
+	/// Dispute information indexed by Parachain-Id; data is DisputeInfo
+	Dispute(u32),
+}
+
+/// A type that defines prefix + hash itself
+pub(crate) type CollectorStorageApi = ApiService<H256, CollectorPrefixType>;
+
+/// A structure used to track disputes progress
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct DisputeInfo {
+	pub initiated: <PolkadotConfig as subxt::Config>::BlockNumber,
+	pub dispute: SubxtDispute,
+	pub parachain_id: u32,
+	pub outcome: Option<SubxtDisputeResult>,
+	pub concluded: Option<<PolkadotConfig as subxt::Config>::BlockNumber>,
+}
+
+/// The current state of the collector, used to detect forks and track candidates
+/// and other events without query storage
+#[derive(Default)]
+struct CollectorState {
+	/// The current block number
+	current_relay_chain_block_number: u32,
+	/// A list of hashes at this height (e.g. if we see forks)
+	current_relay_chain_block_hashes: Vec<H256>,
+	/// A list of candidates seen, indexed by parachain id
+	candidates_seen: BTreeMap<u32, Vec<H256>>,
+	/// A list of disputes seen, indexed by parachain id
+	disputes_seen: BTreeMap<u32, Vec<DisputeInfo>>,
+	/// A current session index
+	current_session_index: u32,
+}
+
+/// Provides collector new head events split by parachain
+#[derive(Clone, Debug)]
+pub struct NewHeadEvent {
+	/// Relay parent block number
+	pub relay_parent_number: <PolkadotConfig as subxt::Config>::BlockNumber,
+	/// Relay parent block hash (or hashes in case of the forks)
+	pub relay_parent_hashes: Vec<H256>,
```

Comment:
I think this shouldn't be a vec as I would expect to see distinct `NewHeadEvent` with same block number.